# IS 362 - Project 4

### Johnny Zgombic

Your assignment is to:  
•Start with the mushroom data in the pandas DataFramethat you constructed in your “Assignment – Preprocessing Data with sci-kit learn.”  
•Use scikit-learn to determine which of the two predictor columns that you selected (odor and one other column of your choice) most accurately predicts whether or not a mushroom is poisonous.  There is an additional challenge here—to use scikit-learn’s predictive classifiers, you’ll want to convert each of your two (numeric categorical) predictor columns into a set of columns.  See for one approach pandas get_dummies() method.  
•Clearly state your conclusions along with any recommendations for further analysis

**The first step is to import all of the necessary tools that we need to complete our tasks.**

In [8]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
sns.set()
import sklearn.model_selection
import sklearn.linear_model
from sklearn import metrics

**The next step is to create our variables, select the columns we will be using, read the data using** *pandas* **and displaying our data using the** *head()* **function.**

In [9]:
url          = 'https://archive.ics.uci.edu/ml/machine-learning-databases/mushroom/agaricus-lepiota.data'
columns      = [0,4,5,22]
column_names = ['Class', 'Bruises', 'Odor', 'Habitat']
mushrooms    = pd.read_csv(url, sep=',', usecols=columns, header=None, names=column_names)
mushrooms.head()

Class Bruises Odor Habitat
0     p       t    p       u
1     e       t    a       g
2     e       t    l       m
3     p       t    p       u
4     e       f    n       g

**Numeric Keys**

Class
    - edible = e = 0
    - poisonous = p = 1 

Bruises
    - bruises = t = 10
    - no = f = 11

Odor
    - almond = a = 20
    - anise = l = 21
    - creosote = c = 22
    - fishy = y = 23
    - foul = f = 24
    - musty = m = 25
    - none = n = 26
    - pungent = p = 27
    - spicy = s = 28

Habitat
    - grasses = g = 30
    - leaves = l = 31
    - meadows = m = 32
    - paths = p = 33
    - urban = u = 34
    - waste = w = 35
    - woods = d = 36

In [10]:
mushrooms['Class']   = mushrooms['Class'].map({'e':0, 'p':1})
mushrooms['Bruises'] = mushrooms['Bruises'].map({'t':10, 'f':11})
mushrooms['Odor']    = mushrooms['Odor'].map({'a':20, 'l':21, 'c':22, 'y':23, 'f':24, 'm':25, 'n':26, 'p':27, 's':28})
mushrooms['Habitat'] = mushrooms['Habitat'].map({'g':30, 'l':31, 'm':32, 'p':33, 'u':34, 'w':35, 'd':36})
mushrooms.head()

Class  Bruises  Odor  Habitat
0      1       10    27       34
1      0       10    20       30
2      0       10    21       32
3      1       10    27       34
4      0       11    26       30

In [11]:
bruises = pd.get_dummies(pd.Series(mushrooms['Bruises']))
odor    = pd.get_dummies(pd.Series(mushrooms['Odor']))
habitat = pd.get_dummies(pd.Series(mushrooms['Habitat']))

In [65]:
mushrooms_new = pd.concat([mushrooms['Class'], bruises, odor, habitat], axis=1)
columns = list(mushrooms_new.iloc[:, :-1])
mushrooms_new.head(10)

Class  10  11  20  21  22  23  24  25  26  27  28  30  31  32  33  34  35  \
0      1   1   0   0   0   0   0   0   0   0   1   0   0   0   0   0   1   0   
1      0   1   0   1   0   0   0   0   0   0   0   0   1   0   0   0   0   0   
2      0   1   0   0   1   0   0   0   0   0   0   0   0   0   1   0   0   0   
3      1   1   0   0   0   0   0   0   0   0   1   0   0   0   0   0   1   0   
4      0   0   1   0   0   0   0   0   0   1   0   0   1   0   0   0   0   0   
5      0   1   0   1   0   0   0   0   0   0   0   0   1   0   0   0   0   0   
6      0   1   0   1   0   0   0   0   0   0   0   0   0   0   1   0   0   0   
7      0   1   0   0   1   0   0   0   0   0   0   0   0   0   1   0   0   0   
8      1   1   0   0   0   0   0   0   0   0   1   0   1   0   0   0   0   0   
9      0   1   0   1   0   0   0   0   0   0   0   0   0   0   1   0   0   0   

   36  
0   0  
1   0  
2   0  
3   0  
4   0  
5   0  
6   0  
7   0  
8   0  
9   0

In [25]:
X = mushrooms_new.iloc[:, :-1].values
Y = mushrooms_new.iloc[:, 1].values

X_train, X_test, Y_train, Y_test = sklearn.model_selection.train_test_split(X, Y, random_state=1)

In [30]:
linreg = sklearn.linear_model.LinearRegression()
linreg.fit(X_train, Y_train)
Y_pred = linreg.predict(X_test)
t = [1, 0]
p = [1, 0]

print(sklearn.metrics.mean_absolute_error(t, p))
print(sklearn.metrics.mean_squared_error(t, p))
print(np.sqrt(sklearn.metrics.mean_squared_error(t, p)))

0.0
0.0
0.0


**We will now run our formula on all the data to see what result we achieve.**

In [64]:
print(sklearn.metrics.mean_absolute_error(Y_test, Y_pred))

6.079058696332566e-15


**We will now remove the** *Habitat* **data to see what result we achieve.**

In [61]:
X = mushrooms_new.iloc[:, :-7].values
Y = mushrooms_new.iloc[:, 1].values

X_train, X_test, Y_train, Y_test = sklearn.model_selection.train_test_split(X, Y, random_state=1)
linreg.fit(X_train, Y_train)
Y_pred = linreg.predict(X_test)

print(sklearn.metrics.mean_absolute_error(Y_test, Y_pred))

1.7709041192276084e-15


**We will now remove the** *Odor* **data to see what result we achieve.**

In [62]:
X = mushrooms_new.iloc[:, 4:12].values
Y = mushrooms_new.iloc[:, 1].values

X_train, X_test, Y_train, Y_test = sklearn.model_selection.train_test_split(X, Y, random_state=1)
linreg.fit(X_train, Y_train)
Y_pred = linreg.predict(X_test)

print(sklearn.metrics.mean_absolute_error(Y_test, Y_pred))

0.26746178207082283


**We will now remove the** *Bruises* **data to see what result we achieve.**

In [63]:
X = mushrooms_new.iloc[:, 2:4].values
Y = mushrooms_new.iloc[:, 1].values

X_train, X_test, Y_train, Y_test = sklearn.model_selection.train_test_split(X, Y, random_state=1)
linreg.fit(X_train, Y_train)
Y_pred = linreg.predict(X_test)

print(sklearn.metrics.mean_absolute_error(Y_test, Y_pred))

6.079058696332566e-15


### Conclusion

**As we can see from the results, the** *Bruises* **category has no effect on predicting anything as when you remove that data, the results are the same**   
   
**When we remove the** *Habitat* **data, we get a much better number at roughly** *1.77*   
   
**We get the best result when we remove the** *Odor* **data. Here we achieved a result of roughly** *.2675*   
   
**What this data tells us is that if we would like to predict which mushrooms are poisonous or not, our best chance would be to look at the** *Habitat* **data we are provided as it is the best indicator as per our results.**